txnkv/transaction/pipelined_flush.go

```diff
@@ -0,0 +1,403 @@
+		Generation:  generation,
+	}
+
+	r := tikvrpc.NewRequest(
+		tikvrpc.CmdFlush, req, kvrpcpb.Context{
+			Priority:               c.priority,
+			SyncLog:                c.syncLog,
+			ResourceGroupTag:       c.resourceGroupTag,
+			DiskFullOpt:            c.txn.diskFullOpt,
+			TxnSource:              c.txn.txnSource,
+			MaxExecutionDurationMs: uint64(client.MaxWriteExecutionTime.Milliseconds()),
+			RequestSource:          c.txn.GetRequestSource(),
+			ResourceControlContext: &kvrpcpb.ResourceControlContext{
+				ResourceGroupName: c.resourceGroupName,
+			},
+		},
+	)
+	return r
+}
+
+func (action actionPipelinedFlush) handleSingleBatch(
+	c *twoPhaseCommitter, bo *retry.Backoffer, batch batchMutations,
+) (err error) {
+	if len(c.primaryKey) == 0 {
+		return errors.New("[pipelined dml] primary key should be set before pipelined flush")
+	}
+
+	tBegin := time.Now()
+	attempts := 0
+
+	req := c.buildPipelinedFlushRequest(batch, action.generation)
+	sender := locate.NewRegionRequestSender(c.store.GetRegionCache(), c.store.GetTiKVClient())
+	var resolvingRecordToken *int
+
+	for {
+		attempts++
+		reqBegin := time.Now()
+		if reqBegin.Sub(tBegin) > slowRequestThreshold {
+			logutil.BgLogger().Warn(
+				"[pipelined dml] slow pipelined flush request",
+				zap.Uint64("startTS", c.startTS),
+				zap.Stringer("region", &batch.region),
+				zap.Int("attempts", attempts),
+			)
+			tBegin = time.Now()
+		}
+		resp, _, err := sender.SendReq(bo, req, batch.region, client.ReadTimeoutShort)
+		// Unexpected error occurs, return it
+		if err != nil {
+			return err
+		}
+		regionErr, err := resp.GetRegionError()
+		if err != nil {
+			return err
+		}
+		if regionErr != nil {
+			// For other region error and the fake region error, backoff because
+			// there's something wrong.
+			// For the real EpochNotMatch error, don't backoff.
+			if regionErr.GetEpochNotMatch() == nil || locate.IsFakeRegionError(regionErr) {
+				err = bo.Backoff(retry.BoRegionMiss, errors.New(regionErr.String()))
+				if err != nil {
+					return err
+				}
+			}
+			if regionErr.GetDiskFull() != nil {
+				storeIds := regionErr.GetDiskFull().GetStoreId()
+				desc := " "
+				for _, i := range storeIds {
+					desc += strconv.FormatUint(i, 10) + " "
+				}
+
+				logutil.Logger(bo.GetCtx()).Error(
+					"Request failed cause of TiKV disk full",
+					zap.String("store_id", desc),
+					zap.String("reason", regionErr.GetDiskFull().GetReason()),
+				)
+
+				return errors.New(regionErr.String())
+			}
+			same, err := batch.relocate(bo, c.store.GetRegionCache())
+			if err != nil {
+				return err
+			}
+			if same {
+				continue
+			}
+			err = c.doActionOnMutations(bo, actionPipelinedFlush{generation: action.generation}, batch.mutations)
+			return err
+		}
+		if resp.Resp == nil {
+			return errors.WithStack(tikverr.ErrBodyMissing)
+		}
+		flushResp := resp.Resp.(*kvrpcpb.FlushResponse)
+		keyErrs := flushResp.GetErrors()
+		var locks []*txnlock.Lock
```

Comment:
We can move the decalaration after the next if, so the else branch is not needed.

Reply:
Done